From 895da13b52779e4c4c8f8c10ab120222f35e9109 Mon Sep 17 00:00:00 2001
From: Punnaiah Choudary Kalluri <punnaiah.choudary.kalluri@xilinx.com>
Date: Wed, 20 Nov 2013 17:18:04 +0530
Subject: [PATCH 046/509] spi: zynq-qspi: Avoid temp variables used in fifo
 read/write operations

https://github.com/analogdevicesinc/linux.git xcomm_zynq
commit 84fa14845c9e4ef61e8ae5c98f516120e8ca5dd4

For better throughput, Avoid temporary variable for storing the fifo data and
copying to user buffer and vice versa, Instead of that use the buffer directly

Signed-off-by: Punnaiah Choudary Kalluri <punnaia@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
Signed-off-by: Zhong Hongbo <hongbo.zhong@windriver.com>
---
 drivers/spi/spi-zynq-qspi.c |   36 ++++++++++++++++++++++++++++--------
 1 files changed, 28 insertions(+), 8 deletions(-)

diff --git a/drivers/spi/spi-zynq-qspi.c b/drivers/spi/spi-zynq-qspi.c
index cadd1b7..15a6d06 100644
--- a/drivers/spi/spi-zynq-qspi.c
+++ b/drivers/spi/spi-zynq-qspi.c
@@ -510,13 +510,21 @@ static int xqspips_setup(struct spi_device *qspi)
  */
 static void xqspips_fill_tx_fifo(struct xqspips *xqspi)
 {
-	u32 data = 0;
 	u32 fifocount;
 
 	for (fifocount = 0; (fifocount < XQSPIPS_FIFO_DEPTH) &&
 			(xqspi->bytes_to_transfer >= 4); fifocount++) {
-		xqspips_copy_write_data(xqspi, &data, 4);
-		xqspips_write(xqspi->regs + XQSPIPS_TXD_00_00_OFFSET, data);
+		if (xqspi->txbuf) {
+			xqspips_write(xqspi->regs + XQSPIPS_TXD_00_00_OFFSET,
+						*((u32 *)xqspi->txbuf));
+			xqspi->txbuf += 4;
+		} else {
+			xqspips_write(xqspi->regs + XQSPIPS_TXD_00_00_OFFSET,
+						0x00);
+		}
+		xqspi->bytes_to_transfer -= 4;
+		if (xqspi->bytes_to_transfer < 0)
+			xqspi->bytes_to_transfer = 0;
 	}
 }
 
@@ -557,13 +565,25 @@ static irqreturn_t xqspips_irq(int irq, void *dev_id)
 		/* Read out the data from the RX FIFO */
 		while (rxcount) {
 
-			data = xqspips_read(xqspi->regs + XQSPIPS_RXD_OFFSET);
-
-			if (xqspi->bytes_to_receive < 4 && !xqspi->is_dual)
+			if (xqspi->bytes_to_receive < 4 && !xqspi->is_dual) {
+				data = xqspips_read(xqspi->regs +
+						XQSPIPS_RXD_OFFSET);
 				xqspips_copy_read_data(xqspi, data,
 					xqspi->bytes_to_receive);
-			else
-				xqspips_copy_read_data(xqspi, data, 4);
+			} else {
+				if (xqspi->rxbuf) {
+					(*(u32 *)xqspi->rxbuf) =
+					xqspips_read(xqspi->regs +
+						XQSPIPS_RXD_OFFSET);
+					xqspi->rxbuf += 4;
+				} else {
+					data = xqspips_read(xqspi->regs +
+							XQSPIPS_RXD_OFFSET);
+				}
+				xqspi->bytes_to_receive -= 4;
+				if (xqspi->bytes_to_receive < 0)
+					xqspi->bytes_to_receive = 0;
+			}
 			rxcount--;
 		}
 
-- 
1.7.5.4

